### PR TITLE
[SYCL] [L0] Make immediate commandlists thread-specific

### PR DESCRIPTION
**Pull request by rdeodhar**
This change makes immediate commandlists used by SYCL queues host-thread specific when SYCL_PI_LEVEL_ZERO_IMMEDIATE_COMMANDLISTS=2 is specified.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1194,17 +1196,27 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
-            ComputeQueueGroup.ZeQueues.size(), CommandListMap.end());
-      }
     } else {
       die("No compute queue available/allowed.");
     }
   }
 
+  if (Device->useImmediateCommandLists()) {
+    auto &ComputeQueueGroupRef = ComputeQueueGroup;
+    bool ThreadSpecificImmCmdLists = Device->useImmediateCommandLists() == 2;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1206]: Please not use hardcoded literal constants in the code, use enums

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 2]
```diff
@@ -882,19 +883,31 @@ struct _pi_queue : _pi_object {
...
 
+  // When using standard commandlists, a group containing compute queue handles.
   pi_queue_group_t ComputeQueueGroup{this, queue_type::Compute};
 
-  // Vector of Level Zero copy command command queue handles.
-  // In this vector, main copy engine, if available, come first followed by
-  // link copy engines, if available.
+  // When using immediate command lists, a set of groups per thread id.
+  // Initialized with calling thread's id when queue is constructed.
+  std::map<std::thread::id, pi_queue_group_t> ComputeQueueGroupsByTID;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:891]: can't the 2 be combined? if not using per-thread lists then there would be a single map entry

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 3]
```diff
@@ -882,19 +883,31 @@ struct _pi_queue : _pi_object {
...
+  // command queue handles. In this vector, main copy engine, if available,
+  // comes first followed by link copy engines, if available.
   pi_queue_group_t CopyQueueGroup{this, queue_type::MainCopy};
 
+  // When using immediate command lists, copy queue groups per thread id
+  // containing a vector of immediate commandlists. In each, the main copy
+  // engine, if available, comes first followed by link copy engines, if
+  // available.
+  // Initialized with calling thread's id when queue is constructed.
+  std::map<std::thread::id, pi_queue_group_t> CopyQueueGroupsByTID;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:903]: please use unordered_map

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -3141,9 +3204,6 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
   }
 
     // intel extensions for GPU information
-  case PI_DEVICE_INFO_DEVICE_ID:
-    return ReturnValue(
-        pi_uint32{Device->ZeDeviceProperties->deviceId});
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3146]: why is this removed?

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 5]
```diff
@@ -841,7 +842,7 @@ struct _pi_queue : _pi_object {
   // PI queue is in general a one to many mapping to L0 native queues.
   struct pi_queue_group_t {
     pi_queue Queue;
-    pi_queue_group_t() = delete;
+    // pi_queue_group_t() = delete;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:845]: why removed? I gauss it's the requirement of std::map but then let's define it so it is not leaving the created object in uninitialized state

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 6]
```diff
@@ -1425,6 +1446,48 @@ pi_result _pi_context::getAvailableCommandList(
...
+
+    bool ThreadSpecificImmCmdLists = Device->useImmediateCommandLists() == 2;
+
+    if (ThreadSpecificImmCmdLists) {
+      // Thread id will be used to create separate imm cmdlists per thread.
+      auto TID = std::this_thread::get_id();
+      if (UseCopyEngine) {
+
+        std::pair<std::map<std::thread::id, pi_queue_group_t>::iterator, bool>
+            Result = CopyQueueGroupsByTID.insert({TID, CopyQueueGroup});
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1460]: What does it mean to use the single "CopyQueueGroup" for each TID?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: OK, done.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 2 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: It is useful to keep the single Compute and Copy queue group members as an initial value. They are set using env var settings for compute and copy engines at queue creation time. Later, as new threads start using the same queue, and new thread-specific Compute/Copy groups are created, the original Compute and Copy queue groups serve as an initializer for the per-thread ones. Otherwise, we would need to do the env var processing again for each new thread. Also, in non-per-thread mode we avoid the map lookup.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 5 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Restored.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1476]: See new comments. The non-threaded Compute/Copy queue groups serve as initializers, (somewhat like OpenMP firstprivate initialization).

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 7]
```diff
@@ -832,7 +832,19 @@ pi_result _pi_device::initialize(int SubSubDeviceOrdinal,
...
+  // One set of compute and copy immediate commandlists per queue.
+  PerQueue,
+  // One set of compute and copy immediate commandlists per host thread that
+  // accesses the queue.
+  PerThreadPerQueue
+};
+
 // Get value of immediate commandlists env var setting or -1 if unset
+// A value of 1 specifies a single set of imm cmdlists per queue.
+// A value of 2 specifies a separate set of imm cmdlists per thread per queue.
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:847]: Please return a value of enum type

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1476]: Sorry, but I still don't understand that. Once you initialize it will stay, but we need different queue groups for different threads.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 8]
```diff
@@ -882,19 +883,34 @@ struct _pi_queue : _pi_object {
...
   pi_queue_group_t ComputeQueueGroup{this, queue_type::Compute};
 
-  // Vector of Level Zero copy command command queue handles.
-  // In this vector, main copy engine, if available, come first followed by
-  // link copy engines, if available.
+  // When using immediate command lists in per-thread mode, the QueueGroups are
+  // per thread, with the thread-id used as a key in this map. The single
+  // QueueGroup above, serves as an initial value to be used to initialize each
+  // new thread's QueueGroup.
+  std::unordered_map<std::thread::id, pi_queue_group_t> ComputeQueueGroupsByTID;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:894]: I think we should just use the map for everything (and have a map of single element for non-per-thread case). Alternatively, if you are worried about map performance for some reason, we should not introduce this new map here, but have this into implementation detail of `pi_queue_group_t`

**Comment by smaslov-intel** [comment]
There are conflicts, please resolve

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 9]
```diff
@@ -1290,25 +1291,19 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
-          Result = ComputeQueueGroupsByTID.insert({TID, ComputeQueueGroup});
-      ComputeQueueGroupRef = Result.first->second;
-    }
-    // Create space to hold immediate commandlists corresponding to the
-    // ZeQueues
-    ComputeQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
-        ComputeQueueGroup.ZeQueues.size(), CommandListMap.end());
-  }
+  // Thread id will be used to create separate queue groups per thread.
+  auto TID = std::this_thread::get_id();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1295]: what if user requested NOT to create lists per-thread ?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 10]
```diff
@@ -1290,25 +1291,19 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
+  std::pair<std::unordered_map<std::thread::id, pi_queue_group_t>::iterator,
+            bool>
+      Result = ComputeQueueGroupsByTID.insert({TID, ComputeQueueGroup});
+  auto &ComputeQueueGroupRef = Result.first->second;
+  // Create space to hold immediate commandlists corresponding to the
+  // ZeQueues
+  ComputeQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
+      ComputeQueueGroup.ZeQueues.size(), CommandListMap.end());
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Yes, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1227]: A single entry in the map is used. All threads use the same immediate commandlists.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1227]: what makes it "single"? what if there are multiple threads in the app, they would have different thread IDs, and this multiple entries in the map, no?

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1227]: See function _pi_queue::getQueueGroup around line 1544. When per-thread is not requested then the first entry in the map is used, without looking at TID. That first entry is created at queue construction time.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1227]: But why create multiple entries here if you are only going to use one?

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 11]
```diff
@@ -371,7 +372,7 @@ struct _pi_device : _pi_object {
   bool ImmCommandListsPreferred;
 
   // Return whether to use immediate commandlists for this device.
-  bool useImmediateCommandLists();
+  int useImmediateCommandLists();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:375]: move https://github.com/intel/llvm/pull/7041/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R765-R773 enum here and return it

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1227]: Multiple entries are not created:

  if (Device->useImmediateCommandLists() == PerThreadPerQueue) {

only if per-thread is requested do we create a new map entry. Otherwise, just the first entry is used.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 11 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.hpp:375]: Ok, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 7 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:802]: Ok, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1239]: but a new map entry is created here: https://github.com/intel/llvm/pull/7041/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R1240

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1239]: There are two calls to "insert". The one in the _pi_queue constructor is called once at queue construction to create an entry using the TID of the thread creating the queue. The only other call to "insert" is in the getQueueGroup function, and that insert is under the if-condition that checks if per-thread-per-queue is set. If per-thread-per-queue is not set, then the map entry created at queue construction is used.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1239]: I see now, thanks. It'd be good to describe this organization of the data structure in the header

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 12]
```diff
@@ -791,14 +782,47 @@ static const size_t ImmCmdListsEventCleanupThreshold = [] {
...
+  switch (ImmediateCommandlistsSetting) {
+  case 0:
+    return NotUsed;
+  case 1:
+    return PerQueue;
+  case 2:
+    return PerThreadPerQueue;
+  default:
+    return NotUsed;
+  }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:825]: please combine these into a single function reading the setting and returning enum

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 13]
```diff
@@ -3676,11 +3760,13 @@ pi_result piextQueueCreate(pi_context Context, pi_device Device,
     // Create as many command-lists as there are queues in the group.
     // With this the underlying round-robin logic would initialize all
     // native queues, and create command-lists and their fences.
-    PI_CALL(warmupQueueGroup(false, Q->ComputeQueueGroup.UpperIndex -
-                                        Q->ComputeQueueGroup.LowerIndex + 1));
+    auto InitialGroup = Q->ComputeQueueGroupsByTID.begin()->second;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3763]: Please add a comment why you are only warming up first queue group here

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 14]
```diff
@@ -3985,7 +4073,15 @@ pi_result piextQueueGetNativeHandle(pi_queue Queue,
 
   // Extract a Level Zero compute queue handle from the given PI queue
   uint32_t QueueGroupOrdinalUnused;
-  *ZeQueue = Queue->ComputeQueueGroup.getZeQueue(&QueueGroupOrdinalUnused);
+  auto TID = std::this_thread::get_id();
+  auto InitialGroup = Queue->ComputeQueueGroupsByTID.begin()->second;
+  std::pair<std::unordered_map<std::thread::id,
+                               _pi_queue::pi_queue_group_t>::iterator,
+            bool>
+      Result = Queue->ComputeQueueGroupsByTID.insert({TID, InitialGroup});
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4081]:
```suggestion
  auto Result = Queue->ComputeQueueGroupsByTID.insert({TID, InitialGroup});
```
A more important Q is why you create a new queue group in the API that is supposed to extract an existing native queue group? (not sure for which TID though) At the very list you should check if a group exists for the current TID and return that, no?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 15]
```diff
@@ -6681,8 +6777,10 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
   std::vector<pi_command_list_ptr_t> CmdLists;
 
   // There must be at least one L0 queue.
-  PI_ASSERT(!Queue->ComputeQueueGroup.ZeQueues.empty() ||
-                !Queue->CopyQueueGroup.ZeQueues.empty(),
+  auto InitialComputeGroup = Queue->ComputeQueueGroupsByTID.begin()->second;
+  auto InitialCopyGroup = Queue->CopyQueueGroupsByTID.begin()->second;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6781]: avoid copies
```suggestion
  auto &InitialComputeGroup = Queue->ComputeQueueGroupsByTID.begin()->second;
  auto &InitialCopyGroup = Queue->CopyQueueGroupsByTID.begin()->second;
```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 16]
```diff
@@ -6681,8 +6777,10 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
...
+  PI_ASSERT(!InitialComputeGroup.ZeQueues.empty() ||
+                !InitialCopyGroup.ZeQueues.empty(),
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6783]: I think you have to count all queues from all threads (the first one may not necessarily have been used)

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6783]: Perhaps you can just move assert after the if/else below where the `CmdLists` computed should be non-empty

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 17]
```diff
@@ -6693,25 +6791,32 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
          It != Queue->CommandListMap.end(); ++It)
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6791]: Please double check that we indeed keep immediate command-lists (from all threads) in this `Queue->CommandListMap`

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6791]: Perhaps we have to do the same as in other places: https://github.com/intel/llvm/pull/7041/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R6976

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 12 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Ok, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 13 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3766]: Done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 14 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: The insert is a conditional insert, if existing entry with same key does not exist.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 15 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Yes, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 16 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6783]: The ASSERT is checking for a valid/usable queue. If there are no ZeQueues of either compute or copy type, the queue is useless.
The presence of commandlists associated with the L0 queues is based on activity on the queue. That is a different check. Here, we do check all groups of all threads. We tolerate no activity on the queue and create a commandlist if the queue had been unused so far.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 17 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6791]: Yes, all commandlists go into the map.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6791]: Then maybe use it everywhere instead of iterating with 2 for-loops?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 16 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6783]: I mean that a queue may be created in one main thread, and then used from other worker threads. In this case the .begin() may not have ant ZeQueues in it. Actually I am not sure if .begin() is even going to point to the main thread in this case.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 18]
```diff
@@ -4075,10 +4078,7 @@ pi_result piextQueueGetNativeHandle(pi_queue Queue,
   uint32_t QueueGroupOrdinalUnused;
   auto TID = std::this_thread::get_id();
   auto InitialGroup = Queue->ComputeQueueGroupsByTID.begin()->second;
-  std::pair<std::unordered_map<std::thread::id,
-                               _pi_queue::pi_queue_group_t>::iterator,
-            bool>
-      Result = Queue->ComputeQueueGroupsByTID.insert({TID, InitialGroup});
+  auto Result = Queue->ComputeQueueGroupsByTID.insert({TID, InitialGroup});
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4081]:
```suggestion
  auto &Result = Queue->ComputeQueueGroupsByTID.insert({TID, InitialGroup});
```

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 14 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Indeed, OK.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 16 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6783]: We are checking if the queue has any associated L0 compute/copy queues. The initial thread may find some and then all subsequent threads will use the same L0 queues. It doesn't matter which entry the "begin" returns. All entries have the same L0 queues. (This is different from commandlists).

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 19]
```diff
@@ -1516,6 +1553,53 @@ pi_result _pi_context::getAvailableCommandList(
...
+      if (!Result.second)
+        return Result.first->second;
+
+      auto &ComputeQueueGroupRef = Result.first->second;
+      // Create space to hold immediate commandlists. They will be created on
+      // demand.
+      ComputeQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
+          InitialGroup.ZeQueues.size(), CommandListMap.end());
+      return ComputeQueueGroupRef;
+    }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1591]: the if/else do the same thing on different objects, please combine

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 18 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Yes, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 17 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6767]: A queue has a map by TID of compute queue groups and a map by TID of copy queue groups. So we have to first iterate over the two groups (compute and copy) and then per-thread entries in each map.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 19 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Yes, even further code reduction was possible and has been done.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 17 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6767]: yes, I understand the reason for 2-deep loopnest, but here we get away by just iterating over `CommandListMap` and assume we see all the same command-lists, right?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 20]
```diff
@@ -1554,50 +1554,25 @@ pi_result _pi_context::getAvailableCommandList(
...
-      // If an entry for this thread exists, use it.
-      if (!Result.second)
-        return Result.first->second;
-
-      auto &ComputeQueueGroupRef = Result.first->second;
-      // Create space to hold immediate commandlists. They will be created on
-      // demand.
-      ComputeQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
+    auto Result = Map.insert({std::this_thread::get_id(), InitialGroup});
+    auto &QueueGroupRef = Result.first->second;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1564]: Please create/adjust the group before inserting it

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 16 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6759]: Hmm, I am now thinking that with immediate command-lists these `ZeQueues` would always be empty

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6759]: When a queue is constructed there are nullptr entries in the computeGroup, totaling the number of CCS's visible. For the copy group, the number of copy engines.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6759]: OK, it will be non-empty vector of null values, right. Not sure what this assert is checking though then
OK. Let it be

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6759]: Yes, the assert is not very useful. It will only be hit if somehow the device presents no compute or copy queues at all, and the first thing done on such a queue is appendBarrier.

And yes, with immediate commandlists being used for the queue, the ZeQueues will contain only nullptrs. Their only purpose when using immediate commandlists is their size, used to allocate that many immediate commandlists.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 20 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1564]: Because "insert" is a conditional insertion, if a new entry is created then the immediate commandlists need initialization.
If we cleared the immediate commandlists in InitialGroup before doing the insert, then the clearing would be wasted if an entry already existed and InitialGroup is not used as an initializer.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 17 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6767]: For standard commandlists there are many commandlists per ZeQueue (L0 queue).
For immediate commandlists there is only one immediate commandlist per L0 queue.
So, for immediate commandlists either type of scan yields the same set of L0 queues.
For standard commandlists if we used CommandListMap we would have to extract the unique L0 queues from the set of commandlists in the map.

You will notice that where ever a two-level for-loop structure is used, the separation between standard/immediate commandlists is inside. In this one case, we check standard/immediate outside, and then do a scan in different ways. This simply matches the code as it was before.
For consistency, this particular instance could also be written as a two-level for-loop with the standard/immediate check inside, if you wish.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6767]: > For consistency, this particular instance could also be written as a two-level for-loop with the standard/immediate check inside, if you wish.

Yes, that's what I was looking for, consistency. Either using one way or another everywhere (and using a single list is probably a faster option). BTW, why do we need to keep the immediate command-lists in the `CommandListMap` at all?
Perhaps, if we could stop doing that then there would not be this confusing diff.

Having said that, this is NIT and should not block this PR.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 20 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1564]: OK

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 17 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: OK, I've changed it for consistency.
The main reason immediate commandlists are also entered into the map is that throughout the plugin, commandlists are passed around as an iterator on that map, never as a commandlist itself. So, to retain that interface I put immediate commandlists in that map too.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: That makes sense, thanks!

**Review by smaslov-intel — APPROVED** [review]

**Review by bader — APPROVED** [review]

File: sycl/doc/EnvironmentVariables.md [hunk 21]
```diff
@@ -244,7 +244,7 @@ variables in production code.</span>
...
-| `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS` | Integer | When set to a positive value enables use of Level Zero immediate commandlists, which means there is no batching and all commands are immediately submitted for execution. Default is 0. Note: When immediate commandlist usage is enabled it is necessary to also set SYCL_PI_LEVEL_ZERO_DEVICE_SCOPE_EVENTS to either 0 or 1.  |
+| `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS` | Integer | When set to a positive value enables use of Level Zero immediate commandlists, which means there is no batching and all commands are immediately submitted for execution. When set to 1, unique immediate commandlists are created for each SYCL queue. When set to 2, unique immediate commandlists are created per host thread per SYCL queue. Default is 0. |
```
Comment (bader) [inline sycl/doc/EnvironmentVariables.md:247]: Do I understand it correctly that "Default is 0" means that immediate command lists are not used?

Comment (rdeodhar) [reply sycl/doc/EnvironmentVariables.md:247]: Yes, by default immediate command lists are not used.
We are working towards making them default.

Comment (againull) [reply sycl/doc/EnvironmentVariables.md:247]: Correct, immediate command lists are not enabled by default yet.
